examples/command_bot.rs

```diff
@@ -0,0 +1,99 @@
+use std::ops::Deref;
+use std::sync::Arc;
+use std::{env, process::exit};
+use url::Url;
+
+use matrix_sdk::{
+    self,
+    events::room::message::{MessageEvent, MessageEventContent, TextMessageEventContent},
+    AsyncClient, AsyncClientConfig, EventEmitter, Room, SyncSettings,
+};
+use tokio::sync::Mutex;
+
+struct CommandBot {
+    client: AsyncClient,
+}
+
+impl CommandBot {
+    pub fn new(client: AsyncClient) -> Self {
+        Self { client }
+    }
+}
+
+#[async_trait::async_trait]
+impl EventEmitter for CommandBot {
+    async fn on_room_message(&mut self, room: Arc<Mutex<Room>>, event: Arc<Mutex<MessageEvent>>) {
+        if let MessageEvent {
+            content: MessageEventContent::Text(TextMessageEventContent { body: msg_body, .. }),
+            ..
+        } = event.lock().await.deref()
+        {
+            let room = room.lock().await;
+            if msg_body.contains("!party") {
+                println!("!party found");
+                let content = MessageEventContent::Text(TextMessageEventContent {
+                    body: "🎉🎊🥳 let's PARTY!! 🥳🎊🎉".to_string(),
+                    format: None,
+                    formatted_body: None,
+                    relates_to: None,
+                });
+                self.client
+                    .room_send(&room.room_id, content, None)
+                    .await
+                    .unwrap();
+                println!("message sent");
+            }
+        }
+    }
+}
+
+#[allow(clippy::for_loop_over_option)]
+async fn login_and_sync(
+    homeserver_url: String,
+    username: String,
+    password: String,
+) -> Result<(), matrix_sdk::Error> {
+    let client_config = AsyncClientConfig::new();
+    // .proxy("http://localhost:8080")?
+    // .disable_ssl_verification();
+    let homeserver_url = Url::parse(&homeserver_url)?;
+    let mut client = AsyncClient::new_with_config(homeserver_url, None, client_config).unwrap();
+
+    client
+        .add_event_emitter(Arc::new(Mutex::new(Box::new(CommandBot::new(
```

Comment:
Not sure you need to put it in a `Arc/Mutex`, you could just derive clone, no?

Reply:
It's in a Arc/Mutex because of this
```rust
pub(crate) async fn emit_timeline_event(&mut self, room_id: &RoomId, event: &mut RoomEvent) {
    match event {
        RoomEvent::RoomMember(mem) => {
            if let Some(ee) = &self.event_emitter { // if we make this
            // &mut self.event_emitter we get immutable borrow inside mut borrow
                if let Some(room) = self.get_room(&room_id) { // this is an &self
                    ee.lock()
                        .await
                        // we need a &mut self here
                        .on_room_member(Arc::clone(&room), Arc::new(Mutex::new(mem.clone())))
                        .await;
                }
            }
        }
```